nix: better dev shell + refactor

Markus328 opened the pull request:
# Issue
With the latest changes (d65fb180ad5ee41049b9d0f9a962901524636b6a) the caelestia plugin was added, and this broke the nix develop environment (no more `qs -p .`, it fails with missing module) since the quickshell package exposed to that environment is not wrapped with `wrapQtAppsHook` and the plugin. 

# Changelist
- fixes this issue wrapping separately the quickshell into a package and adding it to `buildInputs`. 
- use `caelestiaEnv` to automatically add env vars to package and dev shell.
- uneeded binaries were removed from PATH
- removed old `gcc` dependency from caelestia-shell `nativeBuildInputs`

Markus328 commented:
> I just unified the installation method using cmake so this PR won't work anymore. Also, as Quickshell has hot reloading, we don't actually need to have a devshell, we can just use `nix run` to run the shell (no need for wrappers or anything).
> 
> If you would prefer to use a devshell then it would be great if you fixed this PR to work with the new build system. Thanks for the quick fix though!

I would be happy to uptade this PR for the new cmake system.

About the hot reloading, this won't work the same way for nix, since it has `-p $out/share/caelestia/shell.qml` and its in nix store and not writable if not rebuild the whole package. Also, Quickshell doesn't allow to overwrite it after.

soramanew commented:
This seems to work fine for a devshell, idk if it's good practice though
```patch
diff --git a/flake.nix b/flake.nix
index 09a9373..16cb15e 100644
--- a/flake.nix
+++ b/flake.nix
@@ -43,14 +43,17 @@
     });
 
     devShells = forAllSystems (pkgs: {
-      default = let
-        shell = self.packages.${pkgs.system}.caelestia-shell;
-      in
-        pkgs.mkShellNoCC {
-          inputsFrom = [shell];
-          packages = with pkgs; [material-symbols rubik nerd-fonts.caskaydia-cove];
-          CAELESTIA_LIB_DIR = "${shell}/lib";
-        };
+      default = pkgs.mkShell {
+        inputsFrom = [self.packages.${pkgs.system}.caelestia-shell];
+        packages = with pkgs; [material-symbols rubik nerd-fonts.caskaydia-cove];
+        shellHook = ''
+          cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Debug
+          cmake --build build
+          export CAELESTIA_LIB_DIR="$PWD/build/assets/cpp";
+          export QML2_IMPORT_PATH="$PWD/build/qml";
+        '';
+        CAELESTIA_XKB_RULES_PATH = "${pkgs.xkeyboard-config}/share/xkeyboard-config-2/rules/base.lst";
+      };
     });
 
     homeManagerModules.default = import ./nix/hm-module.nix self;
```

This also allows for incremental builds (although if you delete the build dir you'll have to reload the devshell)

soramanew commented:
Why split it into multiple derivations though? Is there a reason to this I'm missing or what, cause you can just build the entire project using cmake instead of splitting it up.

Markus328 commented:
I applied that fix you made, the issue I got earlier was my bad. And I also split the shell derivation into three, to avoid trigger rebuilds in lib and plugin when not changed.

I changed also some `CMakeLists.txt` to allow the split up, anyway that doesn't break the current behavior.

Markus328 commented:
Thanks for the help, bro 😉
